addon/PreviewProvider.js

```diff
@@ -359,15 +359,35 @@ PreviewProvider.prototype = {
   },
 
   /**
-   * Do some pre-processing on the links before inserting them into the metadata
+   * Do some pre-processing on the link before inserting it into the metadata
    * DB and adding them to the metadata cache
    */
-  processAndInsertMetadata(links) {
-    const processedLinks = this._processLinks(links);
-    this.insertMetadata(processedLinks);
+  processAndInsertMetadata(link, metadataSource) {
+    const processedLink = this._processLinks(link);
+    this.insertMetadata(processedLink, metadataSource);
   },
 
   /**
+   * Check if a single link exists in the metadata DB
+   */
+  asyncLinkExist: Task.async(function*(url) {
```

Comment:
I think we should drop this step and this logic here altogether and let metadata store deal with whether it should be written.